eureka2-client/src/main/java/com/netflix/eureka2/client/resolver/RoundRobinServerResolver.java

```diff
@@ -0,0 +1,143 @@
+     * serverCache has warmed up.
+     */
+    private Observable<List<Server>> refreshServerCache() {
+        return serverSource
+                .compose(ChangeNotifications.<Server>buffers())
+                .compose(ChangeNotifications.<Server>snapshots())
+                .materialize()
+                .concatMap(new Func1<Notification<LinkedHashSet<Server>>, Observable<? extends List<Server>>>() {
+                    @Override
+                    public Observable<? extends List<Server>> call(Notification<LinkedHashSet<Server>> rxNotification) {
+                        switch (rxNotification.getKind()) {
+                            case OnNext:
+                                LinkedHashSet<Server> servers = rxNotification.getValue();
+                                if (servers.isEmpty()) {  // if onNext is empty, do nothing and wait
+                                    return Observable.never();
+                                } else {
+                                    logger.info("Populating the serverCache with {} servers", servers.size());
+
+                                    List<Server> newServerCache = new ArrayList<>(servers);
+                                    // do a sort so the list is always roughly in the same order and therefore
+                                    // comparable enough to the previous iteration aside from changes in elements
+                                    Collections.sort(newServerCache);
+                                    serverCacheRef.set(newServerCache);
+                                }
+                                break;
+                            case OnCompleted:  // onCompleted also return the current serverCache
+                                break;
+                            case OnError:
+                                return Observable.error(rxNotification.getThrowable());
+                        }
+                        return Observable.just(serverCacheRef.get());
+                    }
+                })
+                .timeout(warmUpTimeout, timeUnit)
+                .onErrorResumeNext(new Func1<Throwable, Observable<? extends List<Server>>>() {
+                    @Override
+                    public Observable<? extends List<Server>> call(Throwable throwable) {
+                        if (!(throwable instanceof TimeoutException)) {
+                            logger.warn("Exception thrown when connecting the serverCache to the serverSource, using backup values", throwable);
+                        }
+                        return Observable.just(serverCacheRef.get());
+                    }
+                })
+                .take(1);
+    }
+
+    private static Observable<ChangeNotification<Server>> sourceFromList(Server... servers) {
+        return Observable.from(servers)
+                .map(new Func1<Server, ChangeNotification<Server>>() {
```

Comment:
Each subscription will create a new list/change notification objects. As ChangeNotification is not mutable we can create it eagerly or add share operator.
I have seen mapping of T to ChangeNotification<T> in many places. Can we have a generic implementation in ChangeNotifications?